### ISSUE TITLE
RUST_BACKTRACE has no effect on `misaligned pointer dereference` panic?

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#109857]
cc https://github.com/rust-lang/rust/pull/98112 
cc @saethlin 

Code from the pr
````rust
fn main() {
    let mut x = [0u32; 2];
    let ptr: *mut u8 = x.as_mut_ptr().cast::<u8>();
    unsafe {
        *(ptr.add(1).cast::<u32>()) = 42;
    }
}
````

run with `RUST_BACKTRACE=1 ~/.cargo/bin/cargo miri run` seems to not print any backtrace disregarding what the panic output suggests:

````
Preparing a sysroot for Miri (target: x86_64-unknown-linux-gnu)... done
WARNING: Ignoring `RUSTC_WRAPPER` environment variable, Miri does not support wrapping.
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `/home/matthias/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo-miri runner target/miri/x86_64-unknown-linux-gnu/debug/f`
thread 'main' panicked at 'misaligned pointer dereference: address must be a multiple of 0x4 but is 0x278b1', src/main.rs:5:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

````

**Comment by deltragon** [comment]
Does disabling isolation help?
See https://github.com/rust-lang/miri#note-run-with-rust_backtrace1-environment-variable-to-display-a-backtrace

**Comment by saethlin** [comment]
There's no bug here, but I do believe we have a diagnostic issue. @matthiaskrgr is a decently experienced Rust and Miri user, if this is too subtle for Matthias it has probably tripped up new Miri users.

@rust-lang/miri Is there a reasonable way to solve this in Miri? I'm tempted to stick something like this into the standard library right after the `write!` for that `note: run with RUST_BACKTRACE=1`, but that seems like hacking the library because I can't figure out how to detect this situation inside Miri.

**Comment by RalfJung** [comment]
Ah... Yeah I think I was also confused by this at least once.^^ We could consider forwarding that env var by default but then default mode is not fully host-independent any more.

**Comment by saethlin** [comment]
We could also always set it when isolation is enabled. The only potential hazard there is that I have seen one other library check for 

But really what I'm asking is: Can we detect in Miri that we have hit the default panic handler and printed a backtrace, then issue our own little `note: ` after that? My _instinct_ is "no", because that's asking for a lot of details about exactly what library code we executed, but I would like to be sure.

**Comment by oli-obk** [comment]
We could make `default_hook` a `diagnostic_item` and check that in miri. After calling it we could just write another `note: ` to `stderr`, even if writing to `stderr` is not 100% the right thing (there's some thread local shenanigans going on).

**Comment by RalfJung** [comment]
We could also always add some cfg!(miri) code.

**Comment by jyn514** [comment]
Should this be moved to the rust-lang/miri repo?

**Comment by saethlin** [comment]
No, 2 of the 3 proposed solutions involve modifying the standard library code.
